### PR TITLE
feat: support ensure_chunk_conditions_plugin

### LINKED ISSUE #3460: port EnsureChunkConditionsPlugin to rspack
### What problem does this feature solve?

https://github.com/webpack/webpack/blob/main/lib/optimize/EnsureChunkConditionsPlugin.js is important to support external in amd output, related to https://github.com/web-infra-dev/rspack/issues/3457

### What does the proposed API of configuration look like?

same as webpack

### PR DESCRIPTION
**Pull request by suxin2017**
## Related issue (if exists)

resolve #3460 

### <samp>🤖 Generated by Copilot at e4d5853</samp>

This pull request adds chunk condition support to `rspack_core`, a feature that allows external modules to specify whether they should be included in certain chunks or not. It implements a new plugin called `EnsureChunkConditionsPlugin` that ensures that modules with chunk conditions are assigned to the correct chunks based on the condition logic. It also modifies the `Module` trait and its implementations to support chunk conditions.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at e4d5853</samp>

*  Add a new plugin to ensure that modules with chunk conditions are assigned to the appropriate chunks ([link](https://github.com/web-infra-dev/rspack/pull/3502/files?diff=unified&w=0#diff-f2c6e92e3c6a2ce1c6ccff1c0d7deab7656abeb6c7c9696b6d5f82a03f7c116eR1-R78))
   * The plugin iterates over all the modules in the module graph and checks if they have a chunk condition using the `has_chunk_condition` method ([link](https://github.com/web-infra-dev/rspack/pull/3502/files?diff=unified&w=0#diff-6d6ee10fb57dcb86a4a0d44d272155b4466773c75b8a006f4c2b4bbe6209efa0R201-R208))
   * The plugin finds the source chunks and the target chunks for each module with a chunk condition using the `chunk_condition` method ([link](https://github.com/web-infra-dev/rspack/pull/3502/files?diff=unified&w=0#diff-6d6ee10fb57dcb86a4a0d44d272155b4466773c75b8a006f4c2b4bbe6209efa0R201-R208))
*  Modify the `ExternalModule` struct to support chunk conditions ([link](https://github.com/web-infra-dev/rspack/pull/3502/files?diff=unified&w=0#diff-ba05d6e7007157284cbf958d29d0812bedcfceff315bbd95073e99f7f7a1c69eL12-R12), [link](https://github.com/web-infra-dev/rspack/pull/3502/files?diff=unified&w=0#diff-ba05d6e7007157284cbf958d29d0812bedcfceff315bbd95073e99f7f7a1c69eR225-R238))
*  Export the `ensure_chunk_conditions_plugin` module as public in the `lib.rs` file ([link](https://github.com/web-infra-dev/rspack/pull/3502/files?diff=unified&w=0#diff-8d48f40e42d64ce16c665a30278f5e9d4b9426eaa0790c7686c6a787c66d8825R72-R73))

</details>

### CODE REVIEW

File: crates/rspack_core/src/ensure_chunk_conditions_plugin.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
+use std::collections::HashSet;
+
+use rspack_error::Error::InternalError;
+use rspack_identifier::Identifiable;
+
+use crate::{OptimizeChunksArgs, Plugin, PluginContext, PluginOptimizeChunksOutput};
+
+#[derive(Debug)]
+struct EnsureChunkConditionsPlugin;
```
Comment (hyf0) [inline crates/rspack_core/src/ensure_chunk_conditions_plugin.rs:9]: It would be better to move this plugin into a new crate.

Comment (suxin2017) [reply crates/rspack_core/src/ensure_chunk_conditions_plugin.rs:9]: ok

File: crates/rspack_plugin_ensure_chunk_conditions/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  ) -> PluginOptimizeChunksOutput {
+    let compilation = args.compilation;
+
+    let mut source_module_chunks = HashMap::new();
+    compilation
+      .module_graph
+      .modules()
+      .iter()
+      .for_each(|(module_id, module)| {
+        if !module.has_chunk_condition() {
```
Comment (underfin) [inline crates/rspack_plugin_ensure_chunk_conditions/src/lib.rs:24]: Look like here should be `module.has_chunk_condition()`

**Comment by underfin** [comment]
@suxin2017 I will add test for it and rebase it, after this can be merged, what do you think?

**Comment by underfin** [comment]
!canary

**Review by ahabhgk — APPROVED** [review]

**Pull request merged**